### PR TITLE
TSDF tests refactored

### PR DESCRIPTION
**Pull request by savuor**
### Changes
* Duplicated code removal in TSDF tests by implementing them with fixtures and GTest params
  * e.g. separate OCL tests file removed
  * as a result, more test cases are covered
  * the same's done for perf tests

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
> Fix a coordinate problem with Volume::fetchPointsNormals() for TSDF and HashTSDF: they produce points in different coordinate systems
It means that some thresholds should be reduced or be different for the use cases, right?

**Comment by savuor** [comment]
> It means that some thresholds should be reduced or be different for the use cases, right?

No, this is a sign of a bug in `HashTSDF`. I've fixed it.

BTW, it revealed another bug in `HashTSDF` but it's not that critical: it should produce points at 0-surface, in fact it produces points for all non-empty voxels.

**Comment by asmorkalov** [comment]
General comments for tests in the `test_tsdf.cpp`.
- There are `bool onlySemisphere` variables in setup and each test function independently. Should it be common flag?
- There is own `normalize` and `cv::normalize`. It's not clean which one is expected in each call without accurate type checks. I propose to rename own `normalize` to something like `normalizeOrZero` or remove the function and add NaN check to the invocation function.
- Please file a Github ticket for each TODO and HACK and add links to them to the code.

File: modules/3d/test/test_tsdf.cpp [hunk 1]
```diff
@@ -1272,253 +866,293 @@ namespace
...
-}
+    volume->fetchNormals(upoints, unormals);
 
-TEST(TSDF_CPU, valid_points_custom_framesize_frame)
-{
-    OpenCLStatusRevert oclStatus;
-    oclStatus.off();
-    valid_points_test_custom_framesize(VolumeType::TSDF, VolumeTestSrcType::ODOMETRY_FRAME);
-}
+    Mat points, normals, colors;
```
Comment (asmorkalov) [inline modules/3d/test/test_tsdf.cpp:1020]: `colors` is not used in tests.

File: modules/3d/test/test_tsdf.cpp [hunk 2]
```diff
@@ -1272,253 +866,293 @@ namespace
...
+    points  = upoints.getMat(ACCESS_READ);
```
Comment (asmorkalov) [inline modules/3d/test/test_tsdf.cpp:1021]: Points are not checked.

File: modules/3d/test/test_tsdf.cpp [hunk 3]
```diff
@@ -1272,253 +866,293 @@ namespace
...
+    volume->fetchPointsNormals(upoints, unormals);
 
-TEST(HashTSDF_CPU, raycast_common_framesize_normals_mat)
-{
-    OpenCLStatusRevert oclStatus;
-    oclStatus.off();
-    normal_test_custom_framesize(VolumeType::HashTSDF, VolumeTestFunction::RAYCAST, VolumeTestSrcType::MAT);
-}
+    Mat points, normals;
+    points  = upoints.getMat(ACCESS_READ);
```
Comment (asmorkalov) [inline modules/3d/test/test_tsdf.cpp:1039]: Points are not checked.

File: modules/3d/test/test_tsdf.cpp [hunk 4]
```diff
@@ -1272,253 +866,293 @@ namespace
...
+    // TODO: why profile == 2*enface ?
+    float percentValidity = float(enface) / float(profile);
 
-TEST(ColorTSDF_CPU, fetch_normals)
-{
-    OpenCLStatusRevert oclStatus;
-    oclStatus.off();
-    normal_test_custom_framesize(VolumeType::ColorTSDF, VolumeTestFunction::FETCH_NORMALS, VolumeTestSrcType::MAT);
+    ASSERT_GT(profile, 0) << "There are no points in profile";
+    ASSERT_GT(enface, 0) << "There are no points in enface";
```
Comment (asmorkalov) [inline modules/3d/test/test_tsdf.cpp:1122]: It's very weak check for `points`. `colors` part is not checked.

**Comment by asmorkalov** [comment]
Also `normalsCheck` is sanity check that ensures valid normal lenght == 1, but does not check normals direction. It's very weak test for the most of cases.

**Comment by savuor** [comment]
@asmorkalov as for `onlySemisphere` the `valid_points` requires it to be set on while other tests do not care.
This can be changed in the future as soon as we have stronger tests.

I'll create a ticket about making better tests for `Volume` classes.

**Comment by asmorkalov** [comment]
The common approach for perf tests is to run the same function instance with the same parameters several times with `TEST_CYCLE()` macro and get some average statistics. A performance report user get min/max/avg execution time for single function call. The proposed test aggregates time for several rotations together and it's hard to interpret it. I propose to stay with single pose or split the loop on several parameterized test cases, if it makes sense.

**Review by asmorkalov — APPROVED** [review]
:+1:

File: modules/3d/include/opencv2/3d/volume_settings.hpp [hunk 5]
```diff
@@ -31,7 +31,6 @@ class CV_EXPORTS_W VolumeSettings
     * @param volumeType volume type.
     */
     VolumeSettings(VolumeType volumeType);
```
Comment (alalek) [inline modules/3d/include/opencv2/3d/volume_settings.hpp:33]: `explicit` constructors should be used.

Comment (savuor) [reply modules/3d/include/opencv2/3d/volume_settings.hpp:33]: This will be fixed in my next PR

**Pull request merged**